### ISSUE TITLE
SetTag should follow the same convention as LogFields

### ISSUE THREAD

**Issue by MatthewDolan** [opentracing/opentracing-go#186]
## Problem

I noticed that the `log.Field` object borrows from the convention set up by the zap logging library (https://github.com/uber-go/zap) to reduce allocations on logging by having typed functions (e.g. `log.String(...)` or `log.Bool(...)`). The `SetTag(...)` function does not follow that convention though.

## Proposal

I would propose that `SetTag` takes a similar `opentracing.Tag` object that also reduces allocations.

For example:

```
span.SetTag(tag.String("hello", "world"))
```

## Questions to address (if any)

There might need to be some path for rollout and backward compatibility.

**Comment by yurishkuro** [comment]
I think it would've been nice if we did this originally, but at this point it will be a breaking change for SetTag API, which is hard to justify.

**Comment by MatthewDolan** [comment]
I agree that it doesn't quite make sense to make a user in-compatible change here.

Maybe adding a new API `SetTagTyped`. That's a terrible name, but that way it doesn't have impact on consumers. It would just require tracers to upgrade to expose the new API. I feel like changes for tracers are less disruptive than changes for consumers.
